src/daft-logical-plan/src/optimization/rules/reorder_joins/join_graph.rs

```diff
@@ -0,0 +1,734 @@
+                    .filter_map(|e| e.input_mapping().map(|s| (e.name(), s)))
+                    .collect::<HashMap<&'a str, _>>();
+                // To be able to reorder through the current projection, all unresolved columns must either have a
+                // zero-computation projection, or must not be projected by the current Project node (i.e. it should be
+                // resolved from some other branch in the query tree).
+                let reorderable_project = join_conds_to_resolve.iter().all(|(name, _, done)| {
+                    *done
+                        || !schema_names.contains(name)
+                        || projection_input_mapping.contains_key(name.as_str())
+                });
+                if reorderable_project {
+                    for (name, _, done) in join_conds_to_resolve.iter_mut() {
+                        if !*done {
+                            if let Some(new_name) = projection_input_mapping.get(name.as_str()) {
+                                // Remove the current name from the list of schema names so that we can produce
+                                // a set of non-join-key names for the current Project's schema.
+                                schema_names.remove(name);
+                                // If we haven't updated the corresponding entry in the final name map, do so now.
+                                if let Some(final_name) = final_name_map.remove(name) {
+                                    final_name_map.insert(new_name.to_string(), final_name);
+                                }
+                                *name = new_name.to_string();
+                            }
+                        }
+                    }
+                    // Keep track of non-join-key projections so that we can reapply them once we've reordered the query tree.
+                    let non_join_key_projections = projection
+                        .iter()
+                        .filter(|e| schema_names.contains(e.name()))
+                        .cloned()
+                        .collect::<Vec<_>>();
+                    if !non_join_key_projections.is_empty() {
+                        self.final_projections_and_filters
+                            .push(ProjectionOrFilter::Projection(non_join_key_projections));
+                    }
+                    // Continue to children.
+                    self.process_node(input, join_conds_to_resolve, final_name_map);
+                } else {
+                    for (name, _, done) in join_conds_to_resolve.iter_mut() {
+                        if schema_names.contains(name) {
+                            *done = true;
+                        }
+                    }
+                }
+            }
+            LogicalPlan::Filter(Filter {
+                input, predicate, ..
+            }) => {
+                self.final_projections_and_filters
+                    .push(ProjectionOrFilter::Filter(predicate.clone()));
```

Comment:
I believe we also have to rewrite the predicate using the name of the final join output rather than the current name.

See `Expr::replace_columns_with_expressions` which should do this.